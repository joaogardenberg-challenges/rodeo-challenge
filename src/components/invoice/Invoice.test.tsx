import { createContext, useEffect } from 'react'
import { render, screen } from '@testing-library/react'
import { range } from 'lodash'

const useState = jest.fn()
const mockRequest = jest.fn()
jest.doMock('react', () => ({ createContext, useEffect, useState }))
jest.doMock('services/mockRequest', () => mockRequest)

jest.mock('./phase/Phase')
jest.mock('./TotalPrice')

const Invoice = require('./Invoice').default

const phasesCount = 5

const invoice = {
  name: 'Company name',
  phases: range(phasesCount).map((index) => ({ id: index }))
}

describe('Invoice', () => {
  beforeEach(() => {
    mockRequest.mockReturnValue(Promise.resolve({ data: undefined }))
    useState.mockReturnValue([invoice, () => {}])
  })

  it('Renders loading successfully', () => {
    useState.mockReturnValue([undefined, () => {}])
    const { container } = render(<Invoice />)
    const loading = screen.getByTestId('loading')

    expect(mockRequest).toBeCalledTimes(1)
    expect(loading).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  it('Renders the Invoice successfully', () => {
    const { container } = render(<Invoice />)
    const invoiceContainer = screen.getByTestId('invoice')
    const title = screen.getByTestId('title')
    const phases = screen.getAllByTestId('phase')

    expect(mockRequest).toBeCalledTimes(1)
    expect(phases).toHaveLength(phasesCount)
    expect(invoiceContainer).toBeInTheDocument()
    expect(title).toHaveTextContent(`Invoice: ${invoice.name}`)
    expect(container).toMatchSnapshot()
  })
})
