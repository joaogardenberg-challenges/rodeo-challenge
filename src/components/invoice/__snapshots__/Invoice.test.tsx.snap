// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Invoice Renders loading successfully 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin: 10rem 0;
}

<div>
  <main
    class="c0"
    data-testid="loading"
  >
    <p
      class="loading"
    >
      Loading invoice...
    </p>
  </main>
</div>
`;

exports[`Invoice Renders the Invoice successfully 1`] = `
.c1 {
  margin: 0 auto;
  max-width: 100%;
  padding: 1rem;
  width: 60rem;
}

.c1 .page-header {
  padding: 1rem 1rem 2rem;
}

.c1 .page-title {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  font-size: 2.5rem;
  word-break: break-all;
  text-align: center;
}

.c1 .page-title .invoice-name {
  font-size: 0.8em;
}

.c1 .phases {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 3rem;
  padding: 3rem 1rem;
}

.c1 .separator {
  border-bottom: 1px solid #666;
}

.c1 .cost-item {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 1rem;
  padding: 0.5rem;
  -webkit-transition: background-color 200ms ease-in-out;
  transition: background-color 200ms ease-in-out;
}

.c1 .cost-item--labels,
.c1 .cost-item--subtotal,
.c1 .cost-item--total {
  font-weight: 700;
}

.c1 .cost-item--subtotal {
  background-color: #f5f5f5;
}

.c1 .cost-item:not(:last-of-type) {
  border-bottom: 1px solid #ccc;
}

.c0 .cost-item:not(.c0 .cost-item--labels):hover {
  background-color: #eee;
}

.c0 .cost-item:not(.c0 .cost-item--labels):hover.cost-item--subtotal {
  background-color: #e5e5e5;
}

.c0 .cost-item:not(.c0 .cost-item--labels) .cost-item__tax,
.c1 .cost-item:not(.c1 .cost-item--labels) .cost-item__description,
.c1 .cost-item:not(.c1 .cost-item--labels) .cost-item__price,
.c1 .cost-item:not(.c1 .cost-item--labels) .cost-item__number {
  font-family: 'Roboto Mono',monospace;
  font-size: 0.9em;
}

.c1 .cost-item__name {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  word-break: break-all;
}

.c1 .cost-item__tax,
.c1 .cost-item__description,
.c1 .cost-item__price {
  text-align: right;
}

.c1 .cost-item__tax {
  width: 2rem;
}

.c1 .cost-item__description {
  width: 12rem;
}

.c1 .cost-item__price {
  width: 8rem;
}

<div>
  <main
    class="c0 c1"
    data-testid="invoice"
  >
    <header
      class="page-header"
    >
      <h1
        class="page-title"
        data-testid="title"
      >
        <span>
          Invoice
        </span>
        <span
          hidden=""
        >
          : 
        </span>
        <span
          class="invoice-name"
        >
          Company name
        </span>
      </h1>
    </header>
    <div
      class="separator"
    />
    <div
      class="phases"
    >
      <div
        data-testid="phase"
      />
      <div
        data-testid="phase"
      />
      <div
        data-testid="phase"
      />
      <div
        data-testid="phase"
      />
      <div
        data-testid="phase"
      />
    </div>
    <div
      class="separator"
    />
    <div
      data-testid="total-price"
    />
  </main>
</div>
`;
